core/src/main/java/org/infinispan/conflict/impl/StateReceiverImpl.java

```diff
@@ -0,0 +1,233 @@
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.container.DataContainer;
+import org.infinispan.container.entries.InternalCacheEntry;
+import org.infinispan.distribution.LocalizedCacheTopology;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.factories.annotations.Start;
+import org.infinispan.notifications.Listener;
+import org.infinispan.notifications.cachelistener.annotation.DataRehashed;
+import org.infinispan.notifications.cachelistener.event.DataRehashedEvent;
+import org.infinispan.remoting.rpc.RpcManager;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.statetransfer.InboundTransferTask;
+import org.infinispan.statetransfer.StateChunk;
+import org.infinispan.topology.CacheTopology;
+
+/**
+ * @author Ryan Emerson
+ * @since 9.1
+ */
+@Listener
+public class StateReceiverImpl<K, V> implements StateReceiver<K, V> {
+
+   private static Log log = LogFactory.getLog(StateReceiverImpl.class);
+   private static boolean trace = log.isTraceEnabled();
+
+   private String cacheName;
+   private Cache<K, V> cache;
+   private CommandsFactory commandsFactory;
+   private DataContainer<K, V> dataContainer;
+   private RpcManager rpcManager;
+   private long transferTimeout;
+
+   private final ConcurrentHashMap<Integer, SegmentRequest> requestMap = new ConcurrentHashMap<>();
+
+   @Inject
+   public void init(Cache<K, V> cache,
+                    CommandsFactory commandsFactory,
+                    DataContainer<K, V> dataContainer,
+                    RpcManager rpcManager) {
+      this.cache = cache;
+      this.commandsFactory = commandsFactory;
+      this.dataContainer = dataContainer;
+      this.rpcManager = rpcManager;
+   }
+
+   @Start
+   public void start() {
+      this.cache.addListener(this);
+      this.cacheName = cache.getName();
+      this.transferTimeout = cache.getCacheConfiguration().clustering().stateTransfer().timeout();
+   }
+
+   @Override
+   public synchronized void stop() {
+      if (trace) log.tracef("Stop called on StateReceiverImpl for cache %s", cacheName);
+      for (SegmentRequest request : requestMap.values())
+         request.cancel(null);
+   }
+
+   @DataRehashed
+   @SuppressWarnings("unused")
+   synchronized void onDataRehash(DataRehashedEvent dataRehashedEvent) {
+      if (dataRehashedEvent.isPre()) {
+         log.debugf("Cancelling all segment requests as rehash has started %s", dataRehashedEvent);
+
+         for (SegmentRequest request : requestMap.values())
+            request.cancel(new CacheException("Cancelling replica request as the owners of the requested " +
+               "segment have changed."));
+      }
+   }
+
+   @Override
+   public CompletableFuture<List<Map<Address, InternalCacheEntry<K, V>>>> getAllReplicasForSegment(int segmentId, LocalizedCacheTopology topology) {
+      return requestMap.computeIfAbsent(segmentId, id -> new SegmentRequest(id, topology)).requestState();
```

Comment:
I think that if this is called in parallel with `stop()`/`onDataRehash` the request may not be cancelled.